crates/subspace-service/src/piece_cache.rs

```diff
@@ -35,15 +46,56 @@ where
     const KEY_PREFIX: &[u8] = b"piece_cache";
 
     /// Create new instance with specified size (in bytes)
-    pub fn new(aux_store: Arc<AS>, cache_size: u64) -> Self {
+    pub fn new(aux_store: Arc<AS>, cache_size: u64, local_peer_id: PeerId) -> Self {
         let max_pieces_in_cache: PieceIndex = cache_size / PIECE_SIZE as PieceIndex;
+        let local_provided_keys = Self::get_local_provided_keys(aux_store.clone())
+            .expect("DB loading should succeed.")
+            .unwrap_or_default();
+
+        if local_provided_keys.is_empty() {
+            info!("New storage provider initialized.");
+        } else {
+            info!(
+                "Storage provider cache loaded - {} items.",
+                local_provided_keys.len()
+            );
+        }
 
         Self {
             aux_store,
             max_pieces_in_cache,
+            local_peer_id,
+            local_provided_keys,
         }
     }
 
+    fn get_local_provided_keys(
+        aux_store: Arc<AS>,
+    ) -> Result<Option<BTreeSet<PieceIndex>>, Box<dyn Error>> {
+        Ok(aux_store.get_aux(LOCAL_PROVIDED_KEYS)?.map(|data| {
+            let collection: ParityDbKeyCollection =
+                data.try_into().expect("DB loading should succeed.");
+
+            collection.set
+        }))
+    }
+
+    fn put_local_provided_keys(&self) -> Result<(), Box<dyn Error>> {
+        self.aux_store
```

Comment:
There should be TODOs around all such cases, overriding up to 100MiB of data on every insert is pretty bad IMO. I see it is done once for segment, but still, this is quite bad during sync and we need to think how to optimize this soon. In case of node we know that we'll insert pieces, so we can re-insert things later if we detect this upon restart.

Also we can likely avoid insertions of any kind because we know what pieces are inserted already and can simply re-derive local provided keys from that, removing need to write anything at all, at least in this case.